src/stactools/core/utils/raster_footprint.py

```diff
@@ -18,6 +21,622 @@
+
+    Two important operations during this calculation are the densification of
+    the footprint in the native CRS and simplification of the footprint after
+    reprojection to EPSG 4326. If the initial low-vertex polygon in the native
+    CRS is not densified, this can result in a reprojected polygon that does not
+    accurately represent the data footprint. For example, a MODIS asset
+    represented by a rectangular 5 point Polygon in a sinusoidal projection will
+    reproject to a parallelogram in EPSG 4326, when it would be more accurately
+    represented by a polygon with two parallel sides and two curved sides. The
+    difference between these representations is greater the further away from
+    the meridian and equator the asset is located.
+
+    After reprojection to EPSG 4326, the footprint may have more points than
+    desired. This can be simplified to a polygon with fewer points that maintain
+    a maximum distance to the original geometry.
+
+    Args:
+        data_array (numpy.NDArray[Any]): The raster data used for the
+            footprint calculation.
+        crs (CRS): Coordinate reference system of the raster data.
+        transform (Affine): Matrix defining the transformation from pixel to CRS
+            coordinates.
+        precision (int): The number of decimal places to include in the
+            final footprint coordinates.
+        densification_factor (Optional[int]): The factor by which to
+            increase point density within the footprint polygon before
+            projection to EPSG 4326. A factor of 2 would double the density of
+            points (placing one new point between each existing pair of points),
+            a factor of 3 would place two points between each point, etc. Higher
+            densities produce higher fidelity footprints in areas of high
+            projection distortion. Mutually exclusive with
+            ``densification_distance``.
+        densification_distance (Optional[float]): The distance by which to
+            increase point density within the footprint polygon before
+            projection to EPSG 4326. If the distance is set to 2 and the segment
+            length between two polygon vertices is 10, 4 new vertices would be
+            created along the segment. Higher densities produce higher
+            fidelity footprints in areas of high projection distortion.
+            Mutually exclusive with ``densification_factor``.
+        simplify_tolerance (Optional[float]): Distance, in degrees, within which
+            all locations on the simplified polygon will be to the original
+            polygon.
+        no_data (Optional[Union[int, float]]): The nodata value in
+            ``data_array``. If set to None, this will return a footprint
+            including nodata values.
+    """
+
+    crs: CRS
+    """Coordinate reference system of the raster data."""
+
+    data_array: npt.NDArray[Any]
+    """3D array of raster data."""
```

Comment:
Could it be 2D? Or do 2D matrices need to be promoted to 3D?

Reply:
It can be 2D. When initialized, a 2D array will be promoted to 3D since `self.data_mask()` requires a 3D array for `np.sum()` call. I made this a bit more explicit:

- Changed the docstring to "2D or 3D array of raster data."
- Added an assert in `self.data_mask()` that the number of dimension is == 3.